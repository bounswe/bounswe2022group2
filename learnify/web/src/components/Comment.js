import React, { useEffect, useState } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { solid, regular } from '@fortawesome/fontawesome-svg-core/import.macro' // <-- import styles to be used
import MDEditor from "@uiw/react-md-editor";

export default function Comment(props){

  const commentId = props.myComment._id;
  const creator = props.myComment.creator;
  const content = props.myComment.content;
  const images = props.myComment.images;

  const [deleteComment, setDeleteComment] = useState(false);

  const [imageUrl, setImageUrl] = useState("");


  const increaseUp = () => {
      localStorage.setItem((commentId+"upClicked"), true);
      localStorage.setItem(commentId+"upCounter", (1+parseInt(localStorage.getItem(commentId+"upCounter"))));
      localStorage.setItem("commentClicked", true);
      window.location.reload();
  };
  const increaseDown = () => {
      localStorage.setItem((commentId+"downClicked"), true);
      localStorage.setItem(commentId+"downCounter", (1+parseInt(localStorage.getItem(commentId+"downCounter"))));
      localStorage.setItem("commentClicked", true);
      window.location.reload();
  };
  const deleteTheComment = () => {
      setDeleteComment(current => !current);
  };

  const [editComment, setEditComment] = useState(false);

  const editTheComment = () => {
      setValue(content);
      setEditComment(current => !current);
  };

  const [value, setValue] = useState("");

  const token = localStorage.getItem("token");

  const [message, setMessage] = useState("");

  const lsid = props.my_lsid;
  
  const handleSubmitEdit  = () => {
     // editExComment(lsid, commentId, commentTitle, value, imageUrl);
  }
  /*
  const editExComment = async (lsid, commentId, commentTitle, final, imageUrl) => {
      console.log(lsid)
      console.log(commentId)
      console.log(commentTitle)
      console.log(final)
      console.log(imageUrl)
      await fetch(`${process.env.REACT_APP_BACKEND_BASE_URL}learningspace/edit/comment`, {
          method: "PUT",
          body: JSON.stringify({
              ls_id: lsid,
              comment_id: commentId,
              title: commentTitle,
              content: final,
              images: [imageUrl],
          }),
          headers: {
              'Content-type': 'application/json; charset=UTF-8',
              'Authorization': `${token}` , 
          },
      })
          .then((response) => {
              if (response.status === 200) {
                  console.log("successfull")
                  
                  response.json().then( json => {
                      console.log(json.learningSpace.id)
                  });
                  console.log("Learning Space Comment edited successfully!");
                  window.location.reload();
              } else {
                  setMessage("Comment could not be edited!");
                  response.json().then( json => {
                      console.log(json.resultMessage);
                  });
              }
          }
          )
          .catch((error) => {
              console.log(error);
          }
          );
  };
  */

    return(
    <div>
        {!deleteComment &&
        <div className='ls-box-mid'>
            <div className='space-5'></div>
              {images[0] !== "" && images[0] !== null && images[0] !== undefined && <div>
                <img
                    src={images}
                    alt="space"
                    style = {{width: "893.5px",
                    height: "auto",
                    maxWidth: "100%",
                    objectFit: "cover",
                    objectPosition: "center",}}
                />
            <div className='space-5'></div>
            </div>
            }
            <MDEditor.Markdown
                style={{ padding: 15 , overflow: "auto"}}
                source={content}
                linkTarget="_blank"
            />
            <div className='space-5'></div>
            <div className='post-buttons-contaier'>
                <div className='post-box-left'>
                    <div className='ls-button-container2'>
                        <button className='post-upvote-button'>
                            <FontAwesomeIcon icon={solid('caret-up')} color={localStorage.getItem((commentId+"upClicked")) ? "green": "black"} onClick={(localStorage.getItem((commentId+"upClicked")) || localStorage.getItem((commentId+"downClicked"))) ? console.log('onclick..') : increaseUp}/>
                        </button>
                    </div>
                    <div className='post-container-display-item'>
                        <label className="counter__output">{localStorage.getItem(commentId+"upCounter")}</label>
                    </div>
                    <div className='ls-button-container2'>
                        <button className='post-downvote-button'>
                            <FontAwesomeIcon icon={solid('caret-down')} color={localStorage.getItem((commentId+"downClicked")) ? "red": "black"} onClick={(localStorage.getItem((commentId+"upClicked")) || localStorage.getItem((commentId+"downClicked"))) ? console.log('onclick..') : increaseDown}/>
                        </button>
                    </div>
                    <div className='post-container-display-item'>
                        <label className="counter__output">{localStorage.getItem(commentId+"downCounter")}</label>
                    </div>
                    <div className='ls-button-container2'>
                        <div className='post-comment-button'>
                            <a href="/reply">
                                <FontAwesomeIcon icon={regular('comment')} color="black"/>
                            </a>
                        </div>
                    </div>
                    <div className='post-container-display-item'>
                        <label className="counter__output">{0}</label>
                    </div>
                </div>
                <div className='comment-box-right'>
                    <div className='ls-button-container-alt3'>
                        <label className='post-owner-icon-container'>
                            <FontAwesomeIcon icon={solid('user-pen')} color="black"/>
                        </label>
                    </div>
                    <div className='post-owner-display-item'>
                        {creator}
                    </div>
                    <div className='ls-button-container-alt2'>
                        <button className='post-edit-icon-container'>
                                <FontAwesomeIcon icon={solid('edit')} color="black" onClick={editTheComment}/>
                        </button>
                    </div>
                    <div className='ls-button-container-alt2'>
                        <button className='post-delete-button'>
                            <FontAwesomeIcon icon={solid('trash-can')} color="black" onClick={deleteTheComment}/>
                        </button>
                    </div>
                </div>
                </div>
                {editComment && <div>
                <div className='space-3'></div>
                <div className='add-post-box-mid'>
                            <label className="form__label" htmlFor="imageLink">Image Link </label>
                            <div className='space-3'></div>
                            <span className='details-box2' role='textbox' value='nameMessage' contentEditable='true' onInput={(e) => setImageUrl(e.target.textContent)}></span>
                            <div className='space-5'></div>
                            <label className="form__label" htmlFor="postContent">Comment Contents </label>
                            <div className='space-3'></div>
                    <MDEditor height={200} value={value} onChange={setValue} />
                    <div className='space-8'></div>
                    <div>
                      <button className="btn-orange" data-testid="forgotPassword" onClick={() => {handleSubmitEdit()}}>Submit</button>
                    </div>
                    </div>
                </div>}
        </div>
        }
        </div>
    );
}