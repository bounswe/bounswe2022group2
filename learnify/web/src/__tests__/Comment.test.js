import React from "react";
import Comment from "./../components/Comment";
import { screen, render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

const testComment = {
  _id: "1",
  creator: "user1",
  content: "This is a test comment.",
  images: [""],
};

describe("Comment", () => {
  test("should render the comment correctly", async () => {
    const component = renderer
      .create(<BrowserRouter><Comment myComment={testComment} my_lsid="1" /></BrowserRouter>)
      .toJSON();
    expect(component).toMatchSnapshot();
  });

  test("should render the upvote and downvote buttons", async () => {
    render(<BrowserRouter><Comment myComment={testComment} my_lsid="1" /></BrowserRouter>);
    const upvoteButton = screen.getByTestId(/upvote-button/i);
    const downvoteButton = screen.getByTestId(/downvote-button/i);
    expect(upvoteButton).toBeInTheDocument();
    expect(downvoteButton).toBeInTheDocument();
  });

  test("should increase the upvote and downvote counters when clicked", async () => {
    render(<BrowserRouter><Comment myComment={testComment} my_lsid="1" /></BrowserRouter>);
    const upvoteButton = screen.getByTestId(/upvote-button/i);
    const downvoteButton = screen.getByTestId(/downvote-button/i);
    fireEvent.click(upvoteButton);
    fireEvent.click(downvoteButton);
    expect(upvoteButton.textContent).toBe("");
    expect(downvoteButton.textContent).toBe("");
  });

  test("should render the delete button and delete the comment when clicked", async () => {
    render(<BrowserRouter><Comment myComment={testComment} my_lsid="1" /></BrowserRouter>);
    const deleteButton = screen.getByTestId(/delete-button/i);
    fireEvent.click(deleteButton);
    expect(deleteButton).toBeInTheDocument();
    expect(screen.queryByTestId("ls-box-mid")).toBeNull();
  });

  test("should render the edit button and the edit form when clicked", async () => {
    render(<BrowserRouter><Comment myComment={testComment} my_lsid="1" /></BrowserRouter>);
    const editButton = screen.getByTestId(/edit-button/i);
    fireEvent.click(editButton);
    expect(editButton).toBeInTheDocument();
});
});

