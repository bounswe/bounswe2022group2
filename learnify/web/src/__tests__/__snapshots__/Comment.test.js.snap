// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Comment should render the comment correctly 1`] = `
<div>
  <div
    className="ls-box-mid"
  >
    <div
      className="space-5"
    />
    <div
      className="wmde-markdown wmde-markdown-color "
      style={
        Object {
          "overflow": "auto",
          "padding": 15,
        }
      }
    >
      <p>
        This is a test comment.
      </p>
    </div>
    <div
      className="space-5"
    />
    <div
      className="post-buttons-contaier"
    >
      <div
        className="post-box-left"
      >
        <div
          className="ls-button-container2"
        >
          <button
            className="post-upvote-button"
            data-testid="upvote-button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-up "
              color="green"
              data-icon="caret-up"
              data-prefix="fas"
              focusable="false"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M182.6 137.4c-12.5-12.5-32.8-12.5-45.3 0l-128 128c-9.2 9.2-11.9 22.9-6.9 34.9s16.6 19.8 29.6 19.8H288c12.9 0 24.6-7.8 29.6-19.8s2.2-25.7-6.9-34.9l-128-128z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </div>
        <div
          className="post-container-display-item"
        >
          <label
            className="counter__output"
          >
            0
          </label>
        </div>
        <div
          className="ls-button-container2"
        >
          <button
            className="post-downvote-button"
            data-testid="downvote-button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-caret-down "
              color="red"
              data-icon="caret-down"
              data-prefix="fas"
              focusable="false"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 320 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M137.4 374.6c12.5 12.5 32.8 12.5 45.3 0l128-128c9.2-9.2 11.9-22.9 6.9-34.9s-16.6-19.8-29.6-19.8L32 192c-12.9 0-24.6 7.8-29.6 19.8s-2.2 25.7 6.9 34.9l128 128z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </div>
        <div
          className="post-container-display-item"
        >
          <label
            className="counter__output"
          >
            0
          </label>
        </div>
        <div
          className="ls-button-container2"
        >
          <div
            className="post-comment-button"
          >
            <a
              href="/reply"
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-comment "
                color="black"
                data-icon="comment"
                data-prefix="far"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M256 32C114.6 32 .0272 125.1 .0272 240c0 47.63 19.91 91.25 52.91 126.2c-14.88 39.5-45.87 72.88-46.37 73.25c-6.625 7-8.375 17.25-4.625 26C5.818 474.2 14.38 480 24 480c61.5 0 109.1-25.75 139.1-46.25C191.1 442.8 223.3 448 256 448c141.4 0 255.1-93.13 255.1-208S397.4 32 256 32zM256.1 400c-26.75 0-53.12-4.125-78.38-12.12l-22.75-7.125l-19.5 13.75c-14.25 10.12-33.88 21.38-57.5 29c7.375-12.12 14.37-25.75 19.88-40.25l10.62-28l-20.62-21.87C69.82 314.1 48.07 282.2 48.07 240c0-88.25 93.25-160 208-160s208 71.75 208 160S370.8 400 256.1 400z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </a>
          </div>
        </div>
        <div
          className="post-container-display-item"
        >
          <label
            className="counter__output"
          >
            0
          </label>
        </div>
      </div>
      <div
        className="comment-box-right"
      >
        <div
          className="ls-button-container-alt3"
        >
          <label
            className="post-owner-icon-container"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-user-pen "
              color="black"
              data-icon="user-pen"
              data-prefix="fas"
              focusable="false"
              role="img"
              style={Object {}}
              viewBox="0 0 640 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0S96 57.3 96 128s57.3 128 128 128zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512H322.8c-3.1-8.8-3.7-18.4-1.4-27.8l15-60.1c2.8-11.3 8.6-21.5 16.8-29.7l40.3-40.3c-32.1-31-75.7-50.1-123.9-50.1H178.3zm435.5-68.3c-15.6-15.6-40.9-15.6-56.6 0l-29.4 29.4 71 71 29.4-29.4c15.6-15.6 15.6-40.9 0-56.6l-14.4-14.4zM375.9 417c-4.1 4.1-7 9.2-8.4 14.9l-15 60.1c-1.4 5.5 .2 11.2 4.2 15.2s9.7 5.6 15.2 4.2l60.1-15c5.6-1.4 10.8-4.3 14.9-8.4L576.1 358.7l-71-71L375.9 417z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </label>
        </div>
        <div
          className="post-owner-display-item"
        >
          user1
        </div>
        <div
          className="ls-button-container-alt2"
        >
          <button
            className="post-edit-icon-container"
            data-testid="edit-button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-pen-to-square "
              color="black"
              data-icon="pen-to-square"
              data-prefix="fas"
              focusable="false"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 512 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M471.6 21.7c-21.9-21.9-57.3-21.9-79.2 0L362.3 51.7l97.9 97.9 30.1-30.1c21.9-21.9 21.9-57.3 0-79.2L471.6 21.7zm-299.2 220c-6.1 6.1-10.8 13.6-13.5 21.9l-29.6 88.8c-2.9 8.6-.6 18.1 5.8 24.6s15.9 8.7 24.6 5.8l88.8-29.6c8.2-2.8 15.7-7.4 21.9-13.5L437.7 172.3 339.7 74.3 172.4 241.7zM96 64C43 64 0 107 0 160V416c0 53 43 96 96 96H352c53 0 96-43 96-96V320c0-17.7-14.3-32-32-32s-32 14.3-32 32v96c0 17.7-14.3 32-32 32H96c-17.7 0-32-14.3-32-32V160c0-17.7 14.3-32 32-32h96c17.7 0 32-14.3 32-32s-14.3-32-32-32H96z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </div>
        <div
          className="ls-button-container-alt2"
        >
          <button
            className="post-delete-button"
            data-testid="delete-button"
          >
            <svg
              aria-hidden="true"
              className="svg-inline--fa fa-trash-can "
              color="black"
              data-icon="trash-can"
              data-prefix="fas"
              focusable="false"
              onClick={[Function]}
              role="img"
              style={Object {}}
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M135.2 17.7C140.6 6.8 151.7 0 163.8 0H284.2c12.1 0 23.2 6.8 28.6 17.7L320 32h96c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 96 0 81.7 0 64S14.3 32 32 32h96l7.2-14.3zM32 128H416V448c0 35.3-28.7 64-64 64H96c-35.3 0-64-28.7-64-64V128zm96 64c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16zm96 0c-8.8 0-16 7.2-16 16V432c0 8.8 7.2 16 16 16s16-7.2 16-16V208c0-8.8-7.2-16-16-16z"
                fill="currentColor"
                style={Object {}}
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
